File: Commons/Action/src/main/java/com/greentree/commons/action/observable/IntObservable.java
package com.greentree.commons.action.observable;

import java.util.function.IntConsumer;

import com.greentree.commons.action.ListenerCloser;

public interface IntObservable extends RunObservable {
	
	IntObservable NULL = new IntObservable() {
		private static final long serialVersionUID = 1L;

		@Override
		public int listenerSize() {
			return 0;
		}
		
		@Override
		public ListenerCloser addListener(IntConsumer listener) {
			return ListenerCloser.EMPTY;
		}
	};
	
	@Override
	default ListenerCloser addListener(Runnable listener) {
		return addListener(i -> listener.run());
	}
	
	ListenerCloser addListener(IntConsumer listener);
	
}

File: Commons/Action/src/main/java/com/greentree/commons/action/observable/IntObservable.kt
package com.greentree.commons.action.observable

import com.greentree.commons.action.ListenerCloser
import java.util.function.IntConsumer

interface IntObservable : RunObservable {

	override fun addListener(listener: Runnable) = addListener { _ -> listener.run() }

	fun addListener(listener: IntConsumer): ListenerCloser

	data object Empty : IntObservable {

		override val listenerCount: Int
			get() = 0

		override fun addListener(listener: IntConsumer): ListenerCloser {
			return ListenerCloser.Empty
		}

		override fun close() {
		}
	}
}
